File: frontend/src/permissions.ts
// Client-side mirror of the backend Owner / Admin / Member control matrix.
// The single source of truth lives server-side in backend/utils/permissions.py
// (role_of + capability predicates) and the deps.py guards. Hiding a control with
// these predicates is a UX nicety only — the server stays authoritative and every
// rejected call still surfaces the backend's `detail`.

export type ExpenseLike = { created_by?: string | null };
export type TripLike = { admin_ids?: string[] | null };

/**
 * True iff the given user may edit/delete the expense: they created it, or they are a
 * trip admin. A missing `created_by` (legacy rows) makes the creator branch false, so
 * only an admin can modify — matching the backend's safe default. An undefined user
 * (not signed in / still loading) can never modify.
 */
export function canModifyExpense(
  expense: ExpenseLike,
  userId: string | undefined,
  trip: TripLike,
): boolean {
  if (!userId) return false;
  return expense.created_by === userId || (trip.admin_ids ?? []).includes(userId);
}

// ---------------------------------------------------------------------------
// Trip role matrix (mirror of backend/utils/permissions.py)
// ---------------------------------------------------------------------------
export type Role = 'owner' | 'admin' | 'member';
export type RoleTrip = {
  owner_id?: string | null;
  admin_ids?: string[] | null;
  user_ids?: string[] | null;
};

/**
 * Resolve a user's role on a trip: 'owner' (the creator / root admin) supersedes
 * 'admin' (in admin_ids), which supersedes 'member' (in user_ids). Returns null when
 * the user is undefined/loading or not on the trip. Tolerant of missing arrays so a
 * partially-loaded trip never throws.
 */
export function roleOf(trip: RoleTrip, userId: string | undefined): Role | null {
  if (!userId) return null;
  if (trip.owner_id && trip.owner_id === userId) return 'owner';
  if ((trip.admin_ids ?? []).includes(userId)) return 'admin';
  if ((trip.user_ids ?? []).includes(userId)) return 'member';
  return null;
}

// Owner or admin
export function canManageMembers(trip: RoleTrip, userId: string | undefined): boolean {
  const r = roleOf(trip, userId);
  return r === 'owner' || r === 'admin';
}
export function canEditTripSettings(trip: RoleTrip, userId: string | undefined): boolean {
  const r = roleOf(trip, userId);
  return r === 'owner' || r === 'admin';
}

/**
 * True iff the viewer may remove this member row (UX gate). Removal is an admin/owner power, but the
 * owner's own member row (the trip root) is never removable — mirrors the backend, which protects the
 * owner row and otherwise lets admins remove any *settled* member. Settled-ness is data-driven (from
 * the balance engine), so it lives in src/removal.ts, not here.
 */
export function canRemoveMemberRow(
  trip: RoleTrip,
  member: { user_id?: string | null },
  userId: string | undefined,
): boolean {
  if (!canManageMembers(trip, userId)) return false;
  return !(member.user_id && trip.owner_id && member.user_id === trip.owner_id);
}

/**
 * UX mirror of backend can_mark_settlement_paid (Phase 10): a pending settlement may be flipped to
 * paid only by a trip admin/owner OR by the LENDER — the app user linked to the creditor member
 * (settlement.to_member_id). This stops a borrower from self-marking their own debt paid. An
 * undefined user or an unmatched/unlinked creditor member resolves to false. The server stays
 * authoritative; hiding the button is a nicety only.
 */
export function canMarkSettlementPaid(
  trip: RoleTrip,
  settlement: { to_member_id: string },
  userId: string | undefined,
  members: { id: string; user_id?: string | null }[],
): boolean {
  if (!userId) return false;
  const r = roleOf(trip, userId);
  if (r === 'owner' || r === 'admin') return true;
  const lender = members.find((m) => m.id === settlement.to_member_id);
  return !!(lender && lender.user_id && lender.user_id === userId);
}

/**
 * UX mirror of backend can_record_payment (Phase 20): a partial payment along a suggested
 * debtor->creditor pair may be recorded/edited/deleted only by a trip admin/owner OR by the RECEIVER
 * — the app user linked to the creditor member (the pair's to_member_id). The payer can never
 * self-record. An undefined user or an unmatched/unlinked creditor resolves to false. The server
 * stays authoritative; hiding the controls is a nicety only.
 */
export function canRecordPayment(
  trip: RoleTrip,
  toMemberId: string,
  userId: string | undefined,
  members: { id: string; user_id?: string | null }[],
): boolean {
  if (!userId) return false;
  const r = roleOf(trip, userId);
  if (r === 'owner' || r === 'admin') return true;
  const receiver = members.find((m) => m.id === toMemberId);
  return !!(receiver && receiver.user_id && receiver.user_id === userId);
}

// Owner only
export function canManageAdmins(trip: RoleTrip, userId: string | undefined): boolean {
  return roleOf(trip, userId) === 'owner';
}
export function canTransferOwnership(trip: RoleTrip, userId: string | undefined): boolean {
  return roleOf(trip, userId) === 'owner';
}
export function canDeleteTrip(trip: RoleTrip, userId: string | undefined): boolean {
  return roleOf(trip, userId) === 'owner';
}

File: backend/utils/permissions.py
"""Canonical Owner / Admin / Member role layer (Step 23).

Single source of truth for the trip access matrix. These are PURE functions: they
operate on an already-fetched trip dict and never touch the database, so they can be
reused by routes, by the deps.py FastAPI guards, and by unit tests. The DB-touching
guards (``_trip_admin_or_403`` / ``_trip_owner_or_403``) live in utils/deps.py.

The three trip-scoped tiers come straight off the trip document:
  - owner   -> trip["owner_id"]            (the creator / root admin)
  - admin   -> trip["admin_ids"]           (owner is always seeded in here)
  - member  -> trip["user_ids"]            (anyone with access to the trip)

Owner supersedes admin: the owner is also in ``admin_ids`` but ``role_of`` reports
``"owner"`` for them. The application ``super_admin`` is an authenticated, server-verified
identity above those trip roles and does not need to appear in a trip's membership arrays.
"""
from typing import Literal, Optional, Union

from config import SUPER_ADMIN_EMAIL

Role = Literal["super_admin", "owner", "admin", "member"]
Viewer = Optional[Union[str, dict]]


def viewer_id(viewer: Viewer) -> Optional[str]:
    if isinstance(viewer, dict):
        value = viewer.get("id")
        return str(value) if value else None
    return str(viewer) if viewer else None


def is_super_admin(viewer: Viewer) -> bool:
    """Return True only for the server-promoted, exact application operator account."""
    if not isinstance(viewer, dict):
        return False
    email = str(viewer.get("email") or "").strip().lower()
    return email == SUPER_ADMIN_EMAIL and viewer.get("role") == "super_admin"


def role_of(trip: dict, viewer: Viewer) -> Optional[Role]:
    """Return the viewer's effective role on the trip, or None when access is absent.

    Uses ``.get()`` throughout so legacy documents missing ``owner_id`` /
    ``admin_ids`` / ``user_ids`` degrade gracefully instead of raising. A falsy
    viewer id (None / "") is never on a trip.
    """
    if is_super_admin(viewer):
        return "super_admin"
    user_id = viewer_id(viewer)
    if not user_id:
        return None
    if user_id == trip.get("owner_id"):
        return "owner"
    if user_id in trip.get("admin_ids", []):
        return "admin"
    if user_id in trip.get("user_ids", []):
        return "member"
    return None


def can_view(trip: dict, viewer: Viewer) -> bool:
    return role_of(trip, viewer) is not None


def can_manage_members(trip: dict, viewer: Viewer) -> bool:
    return role_of(trip, viewer) in ("super_admin", "owner", "admin")


def can_edit_trip_settings(trip: dict, viewer: Viewer) -> bool:
    return role_of(trip, viewer) in ("super_admin", "owner", "admin")


def can_modify_any_expense(trip: dict, viewer: Viewer) -> bool:
    # creator-or-admin for a *specific* expense stays in deps.can_modify_expense;
    # this is the blanket "may touch any expense" capability of owner/admin.
    return role_of(trip, viewer) in ("super_admin", "owner", "admin")


def can_manage_admins(trip: dict, viewer: Viewer) -> bool:
    return role_of(trip, viewer) in ("super_admin", "owner")


def can_transfer_ownership(trip: dict, viewer: Viewer) -> bool:
    return role_of(trip, viewer) in ("super_admin", "owner")


def can_delete_trip(trip: dict, viewer: Viewer) -> bool:
    return role_of(trip, viewer) in ("super_admin", "owner")


def can_record_payment(trip: dict, to_member_id: Optional[str], viewer: Viewer) -> bool:
    # Phase 20: a payment along a suggested debtor->creditor pair may be recorded/edited/deleted only
    # by a trip admin (owner is always seeded into admin_ids) or by the RECEIVER — the app user linked
    # to the creditor member (to_member_id). The payer can never self-record their own debt as paid.
    # Mirrors can_mark_settlement_paid (Phase 10) but is parametrized on the creditor member id so the
    # POST path (member id from the body) and the PATCH/DELETE path (id from the stored doc) share it.
    if role_of(trip, viewer) in ("super_admin", "owner", "admin"):
        return True
    user_id = viewer_id(viewer)
    receiver = next((m for m in trip.get("members", []) if m["id"] == to_member_id), None)
    if not receiver:
        return False
    if receiver.get("kind") == "family":
        return user_id in (receiver.get("family_member_user_ids") or [])
    return receiver.get("user_id") == user_id
